### PR TITLE
Feature/retry config

### LINKED ISSUE #1215: Package download retry. (conan.io ECONNRESET error)
Hello!

Server sometimes responds with:
('Connection broken: error("(104, \'ECONNRESET\')",)', error("(104, 'ECONNRESET')",)). [Remote: conan.io].

It is reproduced in Travis jobs. E.g. https://travis-ci.org/theirix/conan-libssh2/jobs/222844602 :

```
zlib/1.2.11@lasote/stable: Looking for package 8767e89412729a395e7a478e291eeb99291fe27b in remote 'conan.io' 
Downloading conanmanifest.txt
[==================================================]                  
Downloading conaninfo.txt
[==================================================]                  
Downloading conan_package.tgz
ERROR: Download failed, check server, possibly try again
('Connection broken: error("(104, \'ECONNRESET\')",)', error("(104, 'ECONNRESET')",)). [Remote: conan.io]
Error while executing:
	 conan test_package . -s arch="x86" -s build_type="Release" -s compiler="gcc" -s compiler.version="5.4" -o libssh2:shared="True" -o OpenSSL:no_electric_fence="True" 

```

### PR DESCRIPTION
**Pull request by Minimonium**
Closes #1215

The feature to pull up download/upload `retry` and `retry_wait` configuration to `conan.conf` and environment variables `CONAN_RETRY` and `CONAN_RETRY_WAIT`.

By using them as config variables the user can customize the Conan network behaviour specific to the machine in a unified manner instead of defining the flags in each related command. 
Parameters of the `upload` command and the `tools.download` force override the default values for customizable behaviour.

It requires to change the default values of parameters of the `upload` command to `None`, so we'd fall back to the configuration values.
Also, because of the unified value system default values are now consistent across the API. Because of that, it's required to change old tests since messages on the default behaviour on retries has changed.

The problem of pushing them via an API for the general download (and related) commands is that we'd need to propagate them all across the codebase:
```
uploader_downloader [Downloader]
rest_client_v1, rest_client_v2, rest_client_common? rest_client, net (default is fine)
auth_manager
remote_manager
installer [BinaryInstaller], cmd/download, source [complete_recipe_sources]
manager [install], graph_manager_base [build_graph?]
conan_api [workspace_install, install, install_reference, download]      
command [install, workplace, download]
```
And it would clutter the code with forwarded parameters which is not really worth it.

### CODE REVIEW

**Comment by Minimonium** [comment]
Consider a different name for `retry` and `retry_wait` though

**Comment by lasote** [comment]
I like the idea, but you are changing the default retry attempts, and that is breaking the behavior. Changing that shouldn't be needed, right?

File: conans/client/rest/uploader_downloader.py [hunk 1]
```diff
@@ -254,6 +255,8 @@ def print_progress(output, units, progress=""):
 
 
 def call_with_retry(out, retry, retry_wait, method, *args, **kwargs):
+    if retry is None:
```
Comment (memsharded) [inline conans/client/rest/uploader_downloader.py]: I'd prefer a different approach.
As there are arguments existing, use them, and if another value is to be used from the conan.conf, lets inject it much earlier in the call stack, as soon as possible.
In other words, better to have it coming always from argument than having to read another env-var here.

**Comment by Minimonium** [comment]
> I like the idea, but you are changing the default retry attempts, and that is breaking the behavior. Changing that shouldn't be needed, right?

It's needed to detect when the value is forced and when the value is the defaulted one. I agree though that it's better to preserve existing defaults. Will see how to pull extraction up.

**Comment by Minimonium** [comment]
So, I've pulled up the fallback to the config values for the `upload` in `conan_api` because config values are applied only from that level, but left fall back values in `uploader_downloader` to keep the old behaviour (better to leave.
The only place where I'm not sure it makes sense to have default ones is in `conans/client/cmd/uploader.py` (and we don't have tests to check for that one):
https://github.com/conan-io/conan/pull/5174/files#diff-ad1674306ad35622cdaa15f2eca15c81

Also, I have a problem with the `tools.download`. Since it makes sense to apply the config to everything, but the call here isn't wrapped by config environment variables, so how should one provide the context?
https://github.com/conan-io/conan/blob/develop/conans/client/tools/net.py#L59

**Comment by Minimonium** [comment]
I guess something like #5204? For the tools part.

File: conans/client/conf/__init__.py [hunk 2]
```diff
@@ -375,6 +379,20 @@ def request_timeout(self):
             return float(timeout) if timeout is not None else None
         except ValueError:
             raise ConanException("Specify a numeric parameter for 'request_timeout'")
+              
+    @property
+    def retry(self):
```
Comment (memsharded) [inline conans/client/conf/__init__.py]: These are unused, it seems, unless I am missing something.
They can be removed

File: conans/client/conf/__init__.py [hunk 3]
```diff
@@ -375,6 +379,20 @@ def request_timeout(self):
...
+        try:
+            return self.get_item("general.retry")
+        except ConanException:
+            return None
+
+    @property
+    def retry_wait(self):
```
Comment (memsharded) [inline conans/client/conf/__init__.py]: Unused too

File: conans/client/command.py [hunk 4]
```diff
@@ -22,6 +22,7 @@
 from conans.util.files import exception_message_safe
 from conans.util.files import save
 from conans.util.log import logger
+from conans.util.env_reader import get_env
```
Comment (memsharded) [inline conans/client/command.py]: Unused import?

Comment (Minimonium) [reply conans/client/command.py]: Indeed! Forgot about that one.

```suggestion
```

**Comment by Minimonium** [comment]
Note my comment about `tools.download`.

**Comment by Minimonium** [comment]
Docs: conan-io/docs#1295

**Comment by memsharded** [comment]
Hi @Minimonium 

Each conan_api call is wrapped with:

```python
with tools.environment_append(api._cache.config.env_vars):
                # Patch the globals in tools
                return f(*args, **kwargs)
```
So I'd say tools.download usage of env-vars will be fine.

**Comment by memsharded** [comment]
Merged, will be in next 1.16, many thanks!

Check the first 2 lines I added in your first comment:

> Changelog: Feature: Implement conan.conf retry and retry-wait and CONAN_RETRY and CONAN_RETRY_WAIT to configure all retries for all transfers, including upload, download, and tools.download().
Docs: conan-io/docs#1295


They are used to automatically generated the changelog each release :)

**Pull request merged**